File: Solutions/TwoSum/TwoSum.py
# Justin Thoreson
# Two Sum LeetCode solution

# Latest attempt (best in terms of concise code)
# Runtime: Faster than 83.47% of Python3 online submissions for Two Sum.
# Memory: Less than 42.38% of Python3 online submissions for Two Sum.
# Utilization of a dictionary, similar to the hash map solution in C++
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        d = {};
        for i, k in enumerate(nums):
            if target-k in d:
                return [d[target-k], i]
            else:
                d[k] = i
        
        return [-1, -1]
    
# Original attempt (best in terms of time and space complexity)
# Runtime: Faster than 91.52% of Python3 submissions for Two Sum.
# Memory: Less than 65.67% of Python3 submissions for Two Sum.
# Require duplicate sorted array, 
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        
        tNums = list(nums)
        tNums.sort()
        
        res = [];
        l = 0
        r = len(tNums) - 1

        while l < r:
            sum = tNums[l] + tNums[r]
            if sum < target:
                l += 1
            elif sum > target:
                r -= 1
            elif sum == target:
                index = nums.index(tNums[l])
                res.append(index)
                nums[index] = "Found"         # Account for duplicate entries
                index = nums.index(tNums[r])
                res.append(index)
                return res
        
        return [-1, -1]  

File: Solutions/TwoSum/TwoSum.cpp
// Justin Thoreson
// Two Sum Leetcode Solution

// Best C++ attempt
// Runtime: Faster than 95.35% of C++ online submissions for Two Sum.
// Memory: Less than 24.97% of C++ online submissions for Two Sum.
// Attended a hacking lab on hash maps, solution was learned via group discussion.

// If a number in the vector does not exist in the map, 
// add it as a key with the index as its value.
// Otherwise, if there exists a key k such that the current number + k = target,
// return the index of k and the index of the current number

class Solution {
public:
    vector<int> twoSum(vector<int>& nums, int target) {
        map<int, int> hash;
        vector<int> res;
        for (int i = 0; i < nums.size(); i++) {
            if (hash.find(target-nums[i]) != hash.end())
                return {hash[target-nums[i]], i};
            else
                hash[nums[i]] = i;
        }
        return {-1, -1};
    }
};